src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_ByteString certificate = *(UA_ByteString *)input[0].data;
+    UA_Boolean isTrustedCertificate = *(UA_Boolean *)input[1].data;
+
+    if(!isTrustedCertificate || certificate.length == 0)
+        return UA_STATUSCODE_BADCERTIFICATEINVALID;
+
+    if(server->transaction)
+        return UA_STATUSCODE_BADTRANSACTIONPENDING;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* This Method cannot be called if the containing TrustList Object is open */
+    if(certGroup->fileInfoContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    UA_ByteString certificates[1];
+    certificates[0] = certificate;
+
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCERTIFICATES;
+    trustList.trustedCertificates = certificates;
+    trustList.trustedCertificatesSize = 1;
+
+    return certGroup->addToTrustList(certGroup, &trustList);
+}
+
+static UA_StatusCode
+removeCertificate(UA_Server *server,
+                const UA_NodeId *sessionId, void *sessionHandle,
+                const UA_NodeId *methodId, void *methodContext,
+                const UA_NodeId *objectId, void *objectContext,
+                size_t inputSize, const UA_Variant *input,
+                size_t outputSize, UA_Variant *output) {
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_STRING]) || /*Thumbprint*/
+       !UA_Variant_hasScalarType(&input[1], &UA_TYPES[UA_TYPES_BOOLEAN])) /*IsTrustedCertificate*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_String thumbprint = *(UA_String *)input[0].data;
+    UA_Boolean isTrustedCertificate = *(UA_Boolean *)input[1].data;
+
+    /* TODO: If the Certificate is a CA Certificate that has CRLs then all CRLs for that CA are removed as well */
+
+    if(server->transaction)
+        return UA_STATUSCODE_BADTRANSACTIONPENDING;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* This Method cannot be called if the containing TrustList Object is open */
+    if(certGroup->fileInfoContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_ALL;
+
+    UA_ByteString *certificates;
+    size_t certificatesSize = 0;
+    certGroup->getTrustList(certGroup, &trustList);
+
+    if(isTrustedCertificate) {
+        certificates = trustList.trustedCertificates;
+        certificatesSize = trustList.trustedCertificatesSize;
+    } else {
+        certificates = trustList.issuerCertificates;
+        certificatesSize = trustList.issuerCertificatesSize;
+    }
```

Comment:
You could also take the target addresses here where to finally place the certificate pointer and size information into. This would make the for loop below even smaller.